Allow integrations to adjust user-selected ranges

lyzadanger opened the pull request:
This PR evolved from #5067 (text-range trimming). The behavior changes in this PR ensure that:

* Users cannot annotate empty content (a selection containing only whitespace will never show the Adder)
* The Range used for annotation/highlighting does not contain any leading or trailing whitespace (i.e. is "trimmed"). This change fixes #5024

This PR uses the same implementation of `trimRange` (with some feedback integrated) as #5067, but adds tests. It also adds range trimming at an Integration level via a new `getAnnotatableRange` method (which replaces `canAnnotate`).

Previously, when a user selected (a range of) text, the `guest` would ask its associated Integration whether the Range contained annotatable content by calling `canAnnotate` on the Integration and passing the current `Range` (expecting a boolean result). Now we have a situation in which we want to know both whether the input Range has any annotatable content per the Integration's opinion, and also give the Integration a chance to adjust the `Range` as it sees fit (i.e. current use case: trim it). Naming such an operation is a little awkward; at this point I've got:

```
  /**
   * Return a DOM Range representing the extent of annotatable content within
   * `range`, or `null` if `range` does not contain any annotatable content.
   * `range` may be returned unmodified if already valid.
   */
  getAnnotatableRange(range: Range): Range | null;
```

But am wholly open to other ideas. We could split this into two methods, but I want to avoid performing the range-trimming operation twice for performance reasons.

So, I'm looking for feedback on:

* The set of tests for `trimRange`, and
* The `IntegrationBase` API for validating and optionally trimming user-selected Ranges

## Testing

On this branch, visit http://localhost:3000/document/whitespace

* If you make any selections that only contain whitespace characters of any sort, you should not see the Adder pop up
* If you make a selection that leads or ends with whitespace, the selected range should be unaltered (and you should see the Adder), but if you click "Annotate" or "Highlight", the drawn highlight will snap to/trim to the first and last non-whitespace characters in the selection. The associated annotation or highlight's text content should not have any leading or trailing whitespace.

Contrast this with behavior on the `whitespace-example-document` branch (or `main` if that has been merged):

* If you make any selections that only contain whitespace:
    * If the entire text node is empty, you will not see the Adder, but:
    * If the selection is empty but there is text in the same text node, you will see the Adder, and it is possible to annotate/highlight empty content
* If you make any selections that lead or end in whitespace: That whitespace will be retained as part of the annotated/highlighted content

Fixes #5024

robertknight commented on src/annotator/integrations/pdf.js:
```diff
@@ -198,12 +200,22 @@ export class PDFIntegration extends TinyEmitter {
+        return trimmedRange;
+      }
+    } catch (err) {
+      // no-op
```
What errors are we expecting here, only the `RangeError` if there is no non-whitespace content? I suggest we check for that and re-throw any unexpected errors.

robertknight commented on src/annotator/anchoring/trim-range.ts:
```diff
@@ -0,0 +1,218 @@
+ * non-whitespace character exists between `baseOffset` and the terminus of the
+ * string (start or end depending on `direction`).
+ */
+function trimTextOffset(
```
I think the name of this function and `trimRangeContainers` could be renamed to give better cues about what they returns. `String.prototype.trim()` and `trimRange` return a new trimmed string/range respectively, whereas this function and `trimRangeContainers` return offsets. Maybe something like `findNonSpaceOffset` / `findNonSpaceOffsetInRange`?

lyzadanger replied:
How about `closestNonSpaceOffset`?

lyzadanger replied:
How about `closestNonSpaceOffset` and `closestNonSpacePosition`?

robertknight replied:
Is `closestNonSpacePosition` for searching in the range or an alternative to `closestNonSpaceOffset`? Either of those sound fine for the same-node case.

robertknight commented on src/annotator/anchoring/test/trim-range-test.js:
```diff
@@ -0,0 +1,277 @@
+   * Return a Range that starts before `startElement` and ends after
+   * `endElement`. Use `TextRange` to convert it such that start and end
+   * containers are guaranteed to be text nodes (but not necessarily the correct
+   * ones, per trimming).
```
I would drop the parenthesized comment here, since what is expected/correct depends on the context in which the function is used.

robertknight reviewed (commented):
I took a pass over this and left some feedback. It looks good overall. Regarding VitalSource, I think we should generally make its behavior consistent with the HTML integration unless there is a clear reason to deviate.

lyzadanger commented on src/annotator/integrations/pdf.js:
```diff
@@ -198,12 +200,22 @@ export class PDFIntegration extends TinyEmitter {
-    return canDescribe(range);
+  getAnnotatableRange(range) {
+    try {
+      const trimmedRange = trimRange(TextRange.fromRange(range).toRange());
```
Eureka! Of course. This will be much better.

lyzadanger replied:
Perhaps two static methods on `TextRange` that take and return (DOM) `Range`s:

* `range` (`TextRange.range()`): Equivalent of `TextRange.fromRange(range).toRange()`
* `trimmedRange`: Equivalent of `trimRange(TextRange.fromRange(range).toRange())`

That way only `TextRange` need be dependent on `trimRange`, which itself effectively assumes a `TextRange`-modified `Range`, and other modules using `TextRanges` only need depend on `TextRange`.

Good. Better. Yay.

lyzadanger commented:
@robertknight 

Taking this out of draft now. Meaningful changes since last review:

* Added `trimmedRange` static method to `TextRange` so that integrations do not have a dependency on `trim-range` — nice suggestion!
* VitalSource integration delegates `getAnnotatableRange` to its HTML integration
* PDF and HTML integrations' `getAnnotatableRange` still handle `RangeError`s but will re-throw any other errors
* Test coverage is complete

This PR's diff is now a little chunky and the discussion thread long-ish. Let me know if you'd prefer me to break things up or create a clean PR.

lyzadanger commented on src/annotator/anchoring/test/text-range-test.js:
```diff
@@ -468,5 +452,21 @@ describe('annotator/anchoring/text-range', () => {
+        range.selectNodeContents(el);
+
+        const textRange = TextRange.fromRange(range).toRange();
+        const trimmedRange = TextRange.trimmedRange(range);
```
This test module does not mock `trim-range`. i.e. this is not a unit test. But I'm not sure what the value of a pure unit test would _be_ here; curious as to your thoughts.

lyzadanger commented on src/annotator/anchoring/trim-range.ts:
```diff
@@ -0,0 +1,225 @@
+ *
+ * @throws {RangeError} If no text node with non-whitespace characters found
+ */
+function closestNonSpacePosition(
```
This function returns _position_ metadata, ergo the choice of `closestNonSpacePosition` for its naming. Thoughts?

robertknight replied:
This'll do. An alternative idea might be to name the functions based on what they search in, since that determines the output - eg. `closestNonSpaceInRange` and `closestNonSpaceInString`?

lyzadanger replied:
I am using these latest suggestions, thanks!

lyzadanger commented on src/annotator/integrations/test/html-test.js:
```diff
@@ -67,11 +74,38 @@ describe('HTMLIntegration', () => {
+    let fakeTrimmedRange;
+
+    beforeEach(() => {
+      fakeTrimmedRange = sinon.stub(FakeTextRange, 'trimmedRange');
```
This is not a stubbing strategy I've used before. Does it pass muster or am I overlooking the obvious better way to do this (stubbing a static method on a mocked-imported class)?

robertknight replied:
This looks fine to me. If you mocked a static method on a real import, it would be important to ensure that all such mocks were undone reliably at the end of the test.

robertknight commented on src/annotator/anchoring/text-range.ts:
```diff
@@ -314,4 +316,12 @@ export class TextRange {
+   * whitespace
+   */
+  static trimmedRange(range: Range): Range {
+    return trimRange(TextRange.fromRange(range).toRange());
```
Ah. I was originally thinking to wrap just the `TextRange.fromRange(range).toRange()` part in a static method to explain why this apparently redundant thing was being done.

robertknight commented on src/annotator/anchoring/trim-range.ts:
```diff
@@ -0,0 +1,225 @@
+  };
+
+  // Start the examination at the first text node before/after the
+  // initial boundary node, i.e. do not evaluate the `container` itself
```
Re. the "i.e. ..." at the end, there is no `container` variable here. To restate what is going on, this function is called when the caller knows that the `startContainer` or `endContainer` (depending on direction) is not a suitable start/end point for a trimmed range, and so wants to advance to the next point?

I note that the tests still pass if you remove this `advance` call, so in fact it is just an optimization - unless there is some test case you can think of that would show a difference in behavior? You could simplify the function's behavior conceptually by removing it.

lyzadanger replied:
All righty, I've removed it, and also updated the function's comment as the initial boundary node now _is_ considered in the search.

It's a non-meaningful change relative to this module because this function (newly-named `closestNonSpaceInRange`) is only called by the exported `trimRange` function if it was unable to find a valid offset in the original bounding node via `closestNonSpaceInString`. That is, by the time this function is called, we already "know" that the initial boundary node isn't valid, so there is no behavior change by not `advance()`-ing past it explicitly. If this function were part of the public API, it would represent a behavioral change.

It would be possible to do away with the initial calls to `closestNonSpaceInString` from the `trimRange` function and use only this function (now that it considers boundary nodes). However, using `closestNonSpaceInString` doesn't require node traversal and should be faster.

robertknight commented on src/annotator/anchoring/trim-range.ts:
```diff
@@ -0,0 +1,225 @@
+ *
+ * - `startContainer` and `endContainer` text nodes both contain at least one
+ *   non-whitespace character within the Range's text content
+ * - `startOffset` and `endOffset` both point at non-whitespace characters
```
To clarify, `startOffset` will be just before a non-whitespace character, and `endOffset` will be just after?

lyzadanger replied:
Yep, and I've updated the comment:

```
 * - `startOffset` and `endOffset` both reference non-whitespace characters,
 *   with `startOffset` immediately before the first non-whitespace character
 *   and `endOffset` immediately after the last
 * ```

robertknight reviewed (approved):
I noted a potential simplification of `closestNonSpacePosition` and a few other minor comments. I'll leave you to consider them and get this merged. Good work :)
